File: src/com/gmail/marcosav2010/communicator/packet/wrapper/PacketDecoder.java
package com.gmail.marcosav2010.communicator.packet.wrapper;

import java.io.InputStream;
import java.io.Closeable;
import java.io.EOFException;
import java.io.IOException;
import java.util.UUID;

import com.gmail.marcosav2010.communicator.packet.AbstractPacket;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class PacketDecoder implements Closeable, AutoCloseable {

	@Getter
	private final InputStream in;
	private int length;

	public boolean readBoolean() throws IOException {
		return readByte() == 1;
	}

	public byte readByte() throws IOException {
		addAndCheck(1);
		return (byte) in.read();
	}

	public int readUByte() throws IOException {
		addAndCheck(1);
		return in.read();
	}

	public short readShort() throws IOException {
		return (short) deserialize(2);
	}

	public int readUShort() throws IOException {
		return (int) deserialize(2);
	}

	public int readInt() throws IOException {
		return (int) deserialize(4);
	}

	public long readUInt() throws IOException {
		return deserialize(4);
	}

	public long readLong() throws IOException {
		return deserialize(8);
	}

	public byte[] readBytes() throws IOException {
		int len = readInt();
		return readFully(len);
	}

	public String readString() throws IOException {
		byte[] data = readBytes();
		return new String(data, "UTF-8");
	}

	public UUID readUUID() throws IOException {
		long high = readLong();
		long low = readLong();

		return new UUID(high, low);
	}

	public void readFully(byte[] buffer) throws IOException {
		int r;
		for (int offset = 0; offset < buffer.length; offset += r) {
			r = in.read(buffer, offset, buffer.length - offset);
			if (r < 0)
				throw new EOFException();
		}
	}

	public byte[] readFully(int size) throws IOException {
		addAndCheck(size);
		byte[] b = new byte[size];
		readFully(b);
		return b;
	}

	private long deserialize(int size) throws IOException {
		return deserialize(readFully(size));
	}

	private long deserialize(byte[] data) {
		long ret = 0L;
		for (int i = 0; i < data.length; ++i)
			ret += (data[i] & 0xFFL) << 8 * (data.length - i - 1);
		return ret;
	}

	private void addAndCheck(int i) {
		length += i;
		if (length > AbstractPacket.BASE_SIZE)
			throw new OverExceededByteLimitException(length);
	}

	@Override
	public void close() throws IOException {
		in.close();
	}
}

File: core/src/main/java/com/gmail/marcosav2010/communicator/packet/wrapper/PacketDecoder.kt
package com.gmail.marcosav2010.communicator.packet.wrapper

import com.gmail.marcosav2010.communicator.packet.AbstractPacket
import com.gmail.marcosav2010.communicator.packet.wrapper.exception.OverExceededByteLimitException
import java.io.Closeable
import java.io.EOFException
import java.io.InputStream
import java.nio.charset.StandardCharsets
import java.util.*
import kotlin.experimental.and

class PacketDecoder(private val input: InputStream) : Closeable {

    private var length = 0

    fun readBoolean(): Boolean {
        return readByte().toInt() == 1
    }

    fun readByte(): Byte {
        addAndCheck(1)
        return input.read().toByte()
    }

    fun readUByte(): Int {
        addAndCheck(1)
        return input.read()
    }

    fun readShort(): Short {
        return deserialize(2).toShort()
    }

    fun readUShort(): Int {
        return deserialize(2).toInt()
    }

    fun readInt(): Int {
        return deserialize(4).toInt()
    }

    fun readUInt(): Long {
        return deserialize(4)
    }

    fun readLong(): Long {
        return deserialize(8)
    }

    fun readBytes(): ByteArray {
        val len = readInt()
        return readFully(len)
    }

    fun readString(): String {
        val data = readBytes()
        return String(data, StandardCharsets.UTF_8)
    }

    fun readUUID(): UUID {
        val high = readLong()
        val low = readLong()
        return UUID(high, low)
    }

    private fun readFully(buffer: ByteArray) {
        var r: Int
        var offset = 0
        while (offset < buffer.size) {
            r = input.read(buffer, offset, buffer.size - offset)
            if (r < 0) throw EOFException()
            offset += r
        }
    }

    private fun readFully(size: Int): ByteArray {
        addAndCheck(size)
        val b = ByteArray(size)
        readFully(b)
        return b
    }

    private fun deserialize(size: Int): Long {
        return deserialize(readFully(size))
    }

    private fun deserialize(data: ByteArray): Long {
        var ret = 0L
        for (i in data.indices)
            ret += (data[i] and 0xFFL.toByte()).toInt() shl 8 * (data.size - i - 1)
        return ret
    }

    private fun addAndCheck(i: Int) {
        length += i
        if (length > AbstractPacket.BASE_SIZE) throw OverExceededByteLimitException(length.toLong())
    }

    override fun close() {
        input.close()
    }
}